src/transformStateWithClones.js

```diff
@@ -6,8 +6,32 @@
  *
  * @return {Object[]}
  */
+
 function transformStateWithClones(state, actions) {
-  // write code here
-}
+  let currentState = { ...state };
+  const stateHistory = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        break;
+      case 'removeProperties':
+        currentState = { ...currentState };
```

Comment:
There is no need to create a new copy of currentState here since the object is already being copied in the switch cases when necessary. You can remove this line.